File: corpweb/backend/app/services/antizapret.py
"""
Antizapret configuration service.
Manages /root/antizapret/setup settings and editable config files
via WgBlobStore (stored in DB, not filesystem).
"""
import re
import logging
from typing import Dict, Optional

from sqlalchemy.orm import Session

from app.services.wg_blob_store import WgBlobStore

logger = logging.getLogger(__name__)

ANTIZAPRET_SETUP_FILE = "/root/antizapret/setup"

EDITABLE_FILES: Dict[str, str] = {
    "include_hosts":         "/root/antizapret/config/include-hosts.txt",
    "exclude_hosts":         "/root/antizapret/config/exclude-hosts.txt",
    "include_ips":           "/root/antizapret/config/include-ips.txt",
    "exclude_ips":           "/root/antizapret/config/exclude-ips.txt",
    "allow_ips":             "/root/antizapret/config/allow-ips.txt",
    "forward_ips":           "/root/antizapret/config/forward-ips.txt",
    "include_adblock_hosts": "/root/antizapret/config/include-adblock-hosts.txt",
    "exclude_adblock_hosts": "/root/antizapret/config/exclude-adblock-hosts.txt",
    "remove_hosts":          "/root/antizapret/config/remove-hosts.txt",
}

# Settings stored as y/n
BOOLEAN_SETTINGS = [
    "ROUTE_ALL",
    "DISCORD_INCLUDE",
    "CLOUDFLARE_INCLUDE",
    "AMAZON_INCLUDE",
    "GOOGLE_INCLUDE",
    "WHATSAPP_INCLUDE",
    "TELEGRAM_INCLUDE",
    "HETZNER_INCLUDE",
    "DIGITALOCEAN_INCLUDE",
    "OVH_INCLUDE",
    "AKAMAI_INCLUDE",
    "ROBLOX_INCLUDE",
    "BLOCK_ADS",
    "CLEAR_HOSTS",
    "SSH_PROTECTION",
    "ATTACK_PROTECTION",
    "TORRENT_GUARD",
    "RESTRICT_FORWARD",
    "ALTERNATIVE_CLIENT_IP",
    "ALTERNATIVE_FAKE_IP",
    "CLIENT_ISOLATION",
    "WIREGUARD_BACKUP",
]

# Settings stored as 1/0 (shown as toggle in UI)
NUMERIC_BOOLEAN_SETTINGS = [
    "ANTIZAPRET_DNS",
    "VPN_DNS",
]

STRING_SETTINGS = [
    "WIREGUARD_HOST",
    "WARP_OUTBOUND",
]

ALL_KNOWN_SETTINGS = BOOLEAN_SETTINGS + NUMERIC_BOOLEAN_SETTINGS + STRING_SETTINGS


class AntizapretServiceError(Exception):
    pass


class AntizapretService:

    def __init__(self, db: Session):
        self._store = WgBlobStore(db)

    def get_file_content(self, file_type: str) -> str:
        """Read one of the editable config files from the blob store."""
        if file_type not in EDITABLE_FILES:
            raise AntizapretServiceError(f"Unknown file type: {file_type}")
        path = EDITABLE_FILES[file_type]
        data = self._store.get(path)
        if data is None:
            return ""
        return data.decode("utf-8")

    def save_file_content(self, file_type: str, content: str) -> None:
        """Write one of the editable config files to the blob store."""
        if file_type not in EDITABLE_FILES:
            raise AntizapretServiceError(f"Unknown file type: {file_type}")
        path = EDITABLE_FILES[file_type]
        self._store.put(path, content.encode("utf-8"), by="admin")
        logger.info(f"Saved antizapret file: {path}")

    def get_settings(self) -> Dict[str, Optional[str]]:
        """
        Read /root/antizapret/setup from blob store and return known settings.
        Returns dict where missing keys have value None.
        """
        result: Dict[str, Optional[str]] = {k: None for k in ALL_KNOWN_SETTINGS}
        data = self._store.get(ANTIZAPRET_SETUP_FILE)
        if data is None:
            return result

        content = data.decode("utf-8")
        # Match lines like: KEY=value or KEY="value" or KEY='value'
        pattern = re.compile(r'^([A-Z0-9_]+)=["\']?([^"\'#\n]*)["\']?\s*(?:#.*)?$', re.MULTILINE)
        for match in pattern.finditer(content):
            key = match.group(1)
            value = match.group(2).strip()
            if key in result:
                result[key] = value
        return result

    def update_settings(self, new_settings: Dict[str, str]) -> int:
        """
        Update /root/antizapret/setup in blob store with provided key-value pairs.
        Only keys in ALL_KNOWN_SETTINGS are accepted.
        Preserves comments and unknown lines.
        Returns count of changed parameters.
        """
        data = self._store.get(ANTIZAPRET_SETUP_FILE)
        if data is None:
            raise AntizapretServiceError(f"Setup file not found: {ANTIZAPRET_SETUP_FILE}")

        content = data.decode("utf-8")
        changed = 0

        for key, value in new_settings.items():
            if key not in ALL_KNOWN_SETTINGS:
                continue
            # Sanitize boolean values
            if key in BOOLEAN_SETTINGS:
                value = "y" if value.lower() in ("y", "yes", "true", "1") else "n"
            elif key in NUMERIC_BOOLEAN_SETTINGS:
                value = "1" if value.lower() in ("1", "y", "yes", "true") else "0"

            line_pattern = re.compile(rf'^{re.escape(key)}=.*$', re.MULTILINE)
            new_line = f"{key}={value}"
            if line_pattern.search(content):
                content = line_pattern.sub(new_line, content)
            else:
                content = content.rstrip("\n") + f"\n{new_line}\n"
            changed += 1

        self._store.put(ANTIZAPRET_SETUP_FILE, content.encode("utf-8"), by="admin")
        logger.info(f"Updated {changed} antizapret settings")
        return changed

File: corpweb/frontend/src/api/antizapret.ts
import api from './client'

export type FileType =
  | 'include_hosts'
  | 'exclude_hosts'
  | 'include_ips'
  | 'exclude_ips'
  | 'allow_ips'
  | 'forward_ips'
  | 'include_adblock_hosts'
  | 'exclude_adblock_hosts'
  | 'remove_hosts'

export interface FileContentResponse {
  file_type: string
  content: string
}

export interface AntizapretSettings {
  ROUTE_ALL: string | null
  DISCORD_INCLUDE: string | null
  CLOUDFLARE_INCLUDE: string | null
  AMAZON_INCLUDE: string | null
  GOOGLE_INCLUDE: string | null
  WHATSAPP_INCLUDE: string | null
  TELEGRAM_INCLUDE: string | null
  HETZNER_INCLUDE: string | null
  DIGITALOCEAN_INCLUDE: string | null
  OVH_INCLUDE: string | null
  AKAMAI_INCLUDE: string | null
  ROBLOX_INCLUDE: string | null
  BLOCK_ADS: string | null
  CLEAR_HOSTS: string | null
  WIREGUARD_HOST: string | null
  SSH_PROTECTION: string | null
  ATTACK_PROTECTION: string | null
  TORRENT_GUARD: string | null
  RESTRICT_FORWARD: string | null
  ANTIZAPRET_DNS: string | null
  VPN_DNS: string | null
  ALTERNATIVE_CLIENT_IP: string | null
  ALTERNATIVE_FAKE_IP: string | null
  CLIENT_ISOLATION: string | null
  WARP_OUTBOUND: string | null
  WIREGUARD_BACKUP: string | null
  ESCAPE_ENABLED: string | null
}

export interface DoallResponse {
  status: string
  output: string
  changed?: number
}

export const antizapretApi = {
  getFile: (fileType: FileType) =>
    api.get<FileContentResponse>(`/antizapret/files/${fileType}`).then(r => r.data),

  saveFile: (fileType: FileType, content: string) =>
    api.put<FileContentResponse>(`/antizapret/files/${fileType}`, { content }).then(r => r.data),

  getSettings: () =>
    api.get<AntizapretSettings>('/antizapret/settings').then(r => r.data),

  updateSettings: (settings: Record<string, string>) =>
    api.patch<DoallResponse>('/antizapret/settings', { settings }).then(r => r.data),

  regenerateObfuscation: () =>
    api.post('/antizapret/obfuscation/regenerate').then(r => r.data),
}
